sdks/go/pkg/beam/core/runtime/harness/statecache/statecache.go

```diff
@@ -0,0 +1,172 @@
+			token := string(tok.GetToken())
+			c.setValidToken(transformID, sideInputID, token)
+		}
+	}
+}
+
+// setValidToken adds a new valid token for a request into the SideInputCache struct
+// and maps the transform ID and side input ID pairing to the cache token.
+func (c *SideInputCache) setValidToken(transformID, sideInputID, token string) {
+	idKey := transformID + sideInputID
+	c.idsToTokens[idKey] = token
+	c.validTokens = append(c.validTokens, token)
+}
+
+func (c *SideInputCache) makeAndValidateToken(transformID, sideInputID string) (string, bool) {
+	idKey := transformID + sideInputID
+	// Check if it's a known token
+	tok, ok := c.idsToTokens[idKey]
+	if !ok {
+		return "", false
+	}
+	// Check if the known token is valid for this request
+	for _, t := range c.validTokens {
+		if t == tok {
+			return tok, true
+		}
+	}
+	return "", false
+}
+
+// QueryCache takes a transform ID and side input ID and checking if a corresponding side
+// input has been cached. A query having a bad token (e.g. one that doesn't make a known
+// token or one that makes a known but currently invalid token) is treated the same as a
+// cache miss.
+func (c *SideInputCache) QueryCache(transformID, sideInputID string) *exec.ReusableInput {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	tok, ok := c.makeAndValidateToken(transformID, sideInputID)
+	if !ok {
+		return nil
+	}
+	// Check to see if cached
+	input, ok := c.cache[tok]
+	if !ok {
+		return nil
+	}
+	return input
+}
+
+// SetCache allows a user to place a ReusableInput materialized from the reader into the SideInputCache
+// with its corresponding transform ID and side input ID. If the IDs do not pair with a known, valid token
+// then we silently do not cache the input, as this is an indication that the runner is treating that input
+// as uncacheable.
+func (c *SideInputCache) SetCache(transformID, sideInputID string, input *exec.ReusableInput) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	tok, ok := c.makeAndValidateToken(transformID, sideInputID)
+	if !ok {
+		return nil
+	}
+	if len(c.cache) > c.capacity {
+		err := c.evictElement()
+		if err != nil {
+			return errors.Errorf("Cache at or above capacity, got %v", err)
+		}
+	}
+	c.cache[tok] = input
+	return nil
+}
+
+func (c *SideInputCache) isValid(token string) bool {
+	for _, t := range c.validTokens {
+		if t == token {
+			return true
+		}
+	}
+	return false
+}
+
+// evictElement randomly evicts a ReusableInput that is not currently valid from the cache.
+// It should only be called by a goroutine that obtained the lock in SetCache.
+func (c *SideInputCache) evictElement() error {
+	deleted := false
+	// Select a key from the cache at random
+	for k := range c.cache {
+		// Do not evict an element if it's currently valid
+		if !c.isValid(k) {
+			delete(c.cache, k)
+			deleted = true
+			break
+		}
+	}
+	// Nothing is deleted if every side input is still valid, meaning that the cache size
+	// is likely too small.
+	if !deleted {
```

Comment:
I do like the idea of moving to populating a metrics for this instead of throwing an error.

Reply:
Made a first run at it, hopefully it's what you have in mind. All of the operations that would add metrics (hit, miss, evict, failed evict) are already behind having the lock so that was pretty straightforward.

Reply:
Metrics: Great!.

But also, as mentioned, we probably always want to delete something if an eviction is necessary to avoid memory problems which will just end up crashing a worker if left to grow.

It's totally OK to do it if the currently active set of tokens says everything is being used. Hence the value of the "InUseEvictionCalls" metric. We/users can see if it's a problem, and we can adjust the default/ or make it configurable for users and they can make that choice if needed.